blogposts/2021/how-we-built-batch-changes.md

```diff
@@ -0,0 +1,202 @@
+was called _Automation_ and it was already implemented, kinda: [Quinn][quinn]
+had built an impressive prototype to show to customers and ask them whether
+they'd use something like this. More than one said "yes, I would, and _exactly
+like that_". "[Build one to throw
+away](https://wiki.c2.com/?PlanToThrowOneAway)" is what they say - but do you?
+Do you really throw away a prototype that's more than 10k lines of code and that
+customers pointed at and said "I want this"?
+
+We did, with [this pull request][pr1] — a placeholder here for all of the code
+we wrote in September and October 2019 to rebuild the functionality of the
+prototype from scratch.
+
+What made us do it though? We came to the conclusion that building the product
+from the ground up, thereby fully understanding and owning the code is more
+important to the long-term success of the project and the team than merging a
+prototype to get something out there as fast as possible. From my perspective
+now this was one of the key decisions that helped us avoid accumulating large
+amounts of technical debt, even though we've constantly shipped new things.
+
+Only a few months later, though, we were stuck. We realised that what we had
+built and shipped as an alpha to customers just didn't work on a conceptual
+level. In that version, batch changes (called _campaigns_ at the time) were
+executed on the Sourcegraph instance to produce changes in repositories, which
+was cool from a technical standpoint but it felt clunky and slow, and extending
+which types of change could be made required us to add new code. "Meh" is a good
+word to use here.
+
+So at this point we had thrown the prototype away, built the product from
+scratch and had to realise that, well, _this isn't it_. So what did we do?
+
+[We threw half of it away][pr2]. The trigger was this idea: what if we simply
+don't produce changes on the server-side but instead accept _diffs_ through
+the API and, as part of the Sourcegraph CLI, provide a tool to produce those
+diffs wherever you can run the CLI with whatever Docker container you want?
+
+The server-side would then only have concern itself with diffs: how to
+publish them on code hosts as pull and merge requests, how to sync them, how to
+track them, how to rate limit their creation, etc. We wouldn't have to ship a
+new release every time we came up with a new way to produce diffs. And our
+customers that were most interested in making large-scale changes often had
+custom tooling already. With the CLI producing diffs by running any Docker
+container we could give them a place to plug it in — making it large-scale.
+
+This radical change to how the product worked, was, in my opinion, maybe the
+most important one made in the life of Batch Changes. The mental model and the
+abstraction layers it introduced turned out to be incredibly well-fitting. They
+allowed us to go faster and get a _valuable_ tool into the hands of our
+customers even earlier.
+
```

Comment:
```suggestion
## Pull requests 3 & 4: Still not there... more fine-tuning and user testing
```

Reply:
I used two headings for #3 and #4 (and also realised that PR#4 had the wrong URL)